genie-agent/src/main/java/com/netflix/genie/agent/execution/services/impl/FetchingCacheServiceImpl.java

```diff
@@ -0,0 +1,421 @@
+import java.util.Set;
+
+/**
+ * A cache on local disk that uses URIs as keys and transparently downloads
+ * missing entries. It uses file locks to accomplish the following
+ * <p>
+ * Allows multiple agents reading/writing/deleting in the same cache folder
+ * without corrupting the cache
+ * Downloads new versions of resources once they are available remotely
+ * Deletes the old versions of resources once a new version is successfully downloaded
+ * Recovers from partial downloads of resources in case an agent gets killed in the middle of a download
+ * or a download fails for any other reason
+ * <p>
+ * Does NOT handle concurrency within the same agent (not an issue at the moment)
+ * <p>
+ * Cache structure on local disk
+ * Each resource has a hash to represent it. The version of the resource is the remote last modified
+ * timestamp. For each resource and version, a lock file is created. Each process takes a lock on
+ * this file and then downloads the resource to a data.tmp file. Once the download is complete
+ * data.tmp is atomically moved to data.
+ * For e.g for a hash value of 6d331abc92bc8244bc5d41e2107f303a and last modified = 1525456404
+ * cache entries would like the following
+ * {base_dir}/6d331abc92bc8244bc5d41e2107f303a/1525456404/data
+ * {base_dir}/6d331abc92bc8244bc5d41e2107f303a/1525456404/lock
+ * <p>
+ * Deletion of older versions
+ * Once a version is successfully downloaded, any older versions are deleted as a best effort
+ * TODO:Use shared file lock for reading and exclusive lock for writing to the cache
+ *
+ * @author standon
+ * @since 4.0.0
+ */
+@Service
+@Lazy
+@Slf4j
+@Component
+class FetchingCacheServiceImpl implements FetchingCacheService {
+
+    private static final String LOCK_FILE_NAME = "lock";
+    private static final String DATA_FILE_NAME = "data";
+    private static final String DATA_DOWNLOAD_FILE_NAME = "data.tmp";
+    private static final String DUMMY_FILE_NAME = "_";
+    private final ResourceLoader resourceLoader;
+    private final File cacheDirectory;
+    private final FileLockFactory fileLockFactory;
+    private final TaskExecutor cleanUpTaskExecutor;
+
+    FetchingCacheServiceImpl(
+        final ResourceLoader resourceLoader,
+        final ArgumentDelegates.CacheArguments cacheArguments,
+        final FileLockFactory fileLockFactory,
+        @Qualifier("fetchingCacheServiceCleanUpTaskExecutor") final TaskExecutor cleanUpTaskExecutor
+    ) throws IOException {
+        this.resourceLoader = resourceLoader;
+        this.cacheDirectory = cacheArguments.getCacheDirectory();
+        this.fileLockFactory = fileLockFactory;
+        this.cleanUpTaskExecutor = cleanUpTaskExecutor;
+        createDirectoryStructureIfNotExists(cacheDirectory);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void get(final URI sourceFileUri, final File destinationFile) throws DownloadException, IOException {
+        lookupOrDownload(sourceFileUri, destinationFile);
```

Comment:
The `lookupOrDownload` helper doesn't add anything, could have its body here in `get`